lib/collection/src/update_handler.rs

```diff
@@ -213,42 +220,59 @@ impl UpdateHandler {
-                                    // If is more likely to be the real cause of all further problems
-                                    segs.write().report_optimizer_error(error.clone());
-
-                                    // Error of the optimization can not be handled by API user
-                                    // It is only possible to fix after full restart,
-                                    // so the best available action here is to stop whole
-                                    // optimization thread and log the error
-                                    log::error!("Optimization error: {}", error);
-                                    panic!("Optimization error: {error}");
+                let optimizer = optimizer.clone();
+                let segments = segments.clone();
+                let nsi = nonoptimal_segment_ids.clone();
+                scheduled_segment_ids.extend(&nsi);
+                let callback = callback.clone();
+
+                let handle = spawn_stoppable(
+                    // Stoppable task
+                    {
+                        let segments = segments.clone();
+                        move |stopped| {
+                            match optimizer.as_ref().optimize(segments.clone(), nsi, stopped) {
+                                // Perform some actions when optimization if finished
+                                Ok(result) => {
+                                    callback(result);
+                                    result
                                 }
-                            },
+                                // Handle and report errors
+                                Err(error) => match error {
+                                    CollectionError::Cancelled { description } => {
+                                        log::debug!("Optimization cancelled - {}", description);
+                                        false
+                                    }
+                                    _ => {
+                                        segments.write().report_optimizer_error(error.clone());
+
+                                        // Error of the optimization can not be handled by API user
+                                        // It is only possible to fix after full restart,
+                                        // so the best available action here is to stop whole
+                                        // optimization thread and log the error
+                                        log::error!("Optimization error: {}", error);
+                                        panic!("Optimization error: {error}");
+                                    }
+                                },
+                            }
                         }
-                    }));
-                }
+                    },
+                    // Panic handler
+                    Some(Box::new(move |panic_payload| {
```

Comment:
Pointer: this new callback now gets any panic payload and reports the error on the segment holder.